lib/src/frontend/future_matchers.dart

```diff
@@ -30,24 +34,41 @@ final Matcher completes = const _Completes(null);
 /// [throwsA].
 ///
 /// The [description] parameter is deprecated and shouldn't be used.
+///
+/// This returns an [AsyncMatcher], so [expect] won't complete until the matched
+/// future does.
 Matcher completion(matcher, [@deprecated String description]) =>
     new _Completes(wrapMatcher(matcher));
 
-class _Completes extends Matcher {
+class _Completes extends AsyncMatcher {
   final Matcher _matcher;
 
   const _Completes(this._matcher);
 
-  bool matches(item, Map matchState) {
-    if (item is! Future) return false;
-    Invoker.current.addOutstandingCallback();
+  // Avoid async/await so we synchronously start listening to [item].
+  /*FutureOr<String>*/ matchAsync(item) {
+    if (item is! Future) return "was not a Future";
```

Comment:
It's in the `AsyncMatcher.matchAsync()` docs—is it really worth duplicating in the subclasses?